### PR TITLE
v1.1 - import classes from psr/log 1.1, compatible with psr/log 2.0 and 3.0

### LINKED ISSUE #4: Setup strategy
From my understanding, this project have been created to extract [utility classes that where provided](https://github.com/php-fig/log/tree/1.1.0/Psr/Log/Test) by `psr/log ^1.1` but removed from v2 (https://github.com/php-fig/log/pull/76).

This repo should start with a **v1.0.0** that provides nothing, but require `psr/log ^1.1` which provides `TestLogger` and `LoggerInterfaceTest`.

Next, **v1.1.0** requires `psr/log ^2.0|^3.0` and `php >=8.0.0` and provides the classes that were removed from this versions. Older PHP version don't have to be supported since `psr/log >= 2.0.0` require PHP 8.0 (see [composer.json](https://github.com/php-fig/log/blob/2.0.0/composer.json#L14)).

For a project that uses the test classes, needs support for PHP 7 and allows all versions of `psr/log`, it would have the following requirements:

```
"require": {
    "psr/log": "^1.1|^2.0|3.0"
},
"require-dev": {
    "psr/log-utils": "^1.0"
}
```

Seen https://github.com/php-fig/log-util/pull/2#issuecomment-885454569, ~but they don't provide a migration path and looks outdated~ but my proposition is to not change the class names.

### PR DESCRIPTION
**Pull request by GromNaN**
Fix #4
Follow #5 to create version 1.1.0

- Import classes from `psr/log: 1.1.4`
- Add support for `psr/log: 2.0|3.0`
- Drop support for `php <8.0`

### CODE REVIEW

File: README.md
Comment (Jean85) [inline README.md]: I'm not sure this is a good suggestion. Leaving compatibility for older versions makes the package backward compatible, so I would leave at least `^2.0`

File: README.md
Comment (Jean85) [inline README.md]: I would add a small note saying that having a concrete class support all 3 of them is not so straightforward.

File: README.md
Comment (Jean85) [inline README.md]:
```suggestion
The version of `psr/log-utils` required depends on the version of `psr/log` that you are using:
```

File: README.md
Comment (Jean85) [inline README.md]:
```suggestion
| `^1.1.0`    | `1.0.*`         | Empty package, classes are provided by `psr/log`. |
| `^2.0|^3.0` | `^1.1`          | Imports test classes removed from `psr/log`.          |
```

**Review by Jean85 — CHANGES_REQUESTED** [review]
I've asked for a couple of changes that should go on all branches. This means that probably it would be best to split this PR in 2: one for 1.0, and the other (subsequent) one for 1.1, so it will be easier to review, and I will have a clear cut point for the `1.x` branch.

File: README.md
Comment (GromNaN) [reply README.md]: I'd like to make clear that this is the version that composer decides to install after dependency resolution. Not the versions that are set in the `require` section of `composer.json`.

**Comment by GromNaN** [comment]
I updated the README in #5 (which is for tag `1.0.0`) and rebased this PR.

**Comment by Jean85** [comment]
Please rebase/merge main into this one.

**Comment by GromNaN** [comment]
Rebased, but the release date should be updated.

**Review by Jean85 — APPROVED** [review]
Changes looks good to me. I'll convert this to draft though to avoid accidental merging, I would like to have CC approval first.

**Comment by GromNaN** [comment]
@Seldaek @gmponos do you want to review?

**Comment by Seldaek** [comment]
Seems fine from a cursory look.. Sorry but no time to review more in depth at the moment.

File: src/Test/LoggerInterfaceTest.php
Comment (Crell) [inline src/Test/LoggerInterfaceTest.php:21]:
```suggestion
     * Return an instance of the logger to be tested.
     * 
     * @return LoggerInterface
```

Better self-documentation.

File: src/Test/LoggerInterfaceTest.php
Comment (Crell) [inline src/Test/LoggerInterfaceTest.php:34]: I... don't really understand what I'm supposed to do with this method.

File: src/Test/TestLogger.php
Comment (Crell) [inline src/Test/TestLogger.php]: Instead of `AbstractLogger` use `LoggerTrait` directly.  That's all `AbstractLogger` does now.  May as well skip a step.

File: src/Test/TestLogger.php
Comment (Crell) [inline src/Test/TestLogger.php:49]: If these are all testing methods, shouldn't they be `assert*()` type methods?

Also, checking for a record message and a full record array are different operations, so should have separate (properly typed) methods.

File: src/Test/TestLogger.php
Comment (Crell) [inline src/Test/TestLogger.php]: Can be short lambda.

File: src/Test/TestLogger.php
Comment (Crell) [inline src/Test/TestLogger.php]: Can be short lambda.

File: src/Test/TestLogger.php
Comment (Crell) [inline src/Test/TestLogger.php:108]: What is the purpose of this method?  I don't understand why it is here.

File: src/Test/LoggerInterfaceTest.php
Comment (GromNaN) [reply src/Test/LoggerInterfaceTest.php:34]: The logged messages.

This is the main limitation of this class: we need an ability to read what is written (example: read the file.log)

Actual implementation:
https://github.com/Seldaek/monolog/blob/1.26.1/tests/Monolog/PsrLogCompatTest.php#L35-L46

File: src/Test/TestLogger.php
Comment (GromNaN) [reply src/Test/TestLogger.php:49]: This methods doesn't assert anything, but describe the state of the logged messages.
I can't change them for compatibility with `psr/log: 1.1`

File: src/Test/TestLogger.php
Comment (GromNaN) [reply src/Test/TestLogger.php:108]: It provides all the methods described in the header (`hasRecord*` for all log levels).

**Review by GromNaN — COMMENTED** [review]
Thanks for the review.

I took the classes as this, from `psr/log: 1.1.4`. Deliberately limiting changes to avoid confusion. But sure, code can be modernized and improved.

File: src/Test/TestLogger.php
Comment (Crell) [reply src/Test/TestLogger.php:49]: Make a note for future versions then. :-)

File: src/Test/TestLogger.php
Comment (Crell) [reply src/Test/TestLogger.php:108]: I... genuinely hate this approach.  It's harder to understand, harder to debug, and slower.  It has almost no benefit over just writing the methods directly and being done with it.

Comment (GromNaN) [reply src/Test/TestLogger.php:108]: I materialized all the magic methods and deprecated `__call` in https://github.com/php-fig/log-util/pull/6/commits/2c9fe51d0b6133421a0ab05fbb770e7f99bf86bc

**Comment by GromNaN** [comment]
@Jean85 can we move forward?

**Comment by Jean85** [comment]
Code looks fine to me, but I would wait befor merging because we would need to release the 1.0 first.

In any case, we should ask the Core Committee to review and approve all of this, but CC elections are in progress: https://groups.google.com/g/php-fig/c/GdFo9Nv88Jo

I would suggest you to start wrapping up all your work and the issues where we discussed the strategy and present it to the ML; then, we could put it to a vote as the first task for the newly elected CC.

**Comment by javer** [comment]
Are we going to merge it and release somewhen? Because there is no package `psr/log-util` in packagist, there are no any releases in this repo, and this PR is not merged yet. So it's impossible to test services which use Psr\Logger since upgrading to `psr/log` v2.

**Comment by ralflang** [comment]
Merge & release will happen tonight. Have reviewed over the weekend and looks fine.

**Comment by GromNaN** [comment]
Hi, thanks for the [positive votes](https://groups.google.com/g/php-fig/c/OYqyIJIBWGw). Our work is not lost.

Just a quick reminder that there is 2 tags to create after merging:
- `v1.0.0` on the current `main` branch (commit https://github.com/php-fig/log-util/commit/60c112a3f9bcf93abb208917fe6dff4f65959bac)
- `v1.1.0` on this MR merge commit.

Versions are described in https://github.com/php-fig/log-util#versions

File: CHANGELOG.md
Comment (GromNaN) [inline CHANGELOG.md:11]: @ralflang the release date should be updated.

**Comment by Crell** [comment]
The 1.0.0 release will also require another CC vote.  All majors are subject to CC vote.

**Comment by GromNaN** [comment]
Hi @Crell, How can we open the vote for the release?

**Comment by Crell** [comment]
We have a maintainer appointed now, Ralf Lang.  He's on vacation for a week but he can open the vote on the list as soon as he's comfortable with where it is.

**Comment by ralflang** [comment]
Sorry guys I am back and trying to get this done :)

**Comment by derrabus** [comment]
I hate to be that guy, but what stops us from merging this fine PR?

**Comment by GromNaN** [comment]
If anybody is watching this PR. The package was renamed [`fig/log-test`](https://github.com/php-fig/log-test/).

I think this GitHub project should be archived.

**Comment by Jean85** [comment]
I would prefer not to archive for now, since we could need to release patches (i.e. if a future PHP 8.x version breaks anything).

**Pull request closed without merging**
